vscode/src/editor/utils.ts

```diff
@@ -0,0 +1,87 @@
+import * as vscode from 'vscode'
+
+/**
+ * Gets the folding ranges for the given document URI.
+ *
+ * @param uri - The URI of the document to get folding ranges for
+ * @param type - Optional type of folding range to filter on:
+ *   - 'imports' - Only import folding ranges
+ *   - 'comment' - Only comment folding ranges
+ *   - 'regions' - All non-import and non-comment folding ranges
+ * If no type specified, returns all folding ranges.
+ *
+ * @returns The array of folding ranges for the document.
+ */
+export async function getFoldingRange(uri: vscode.Uri, type?: string): Promise<vscode.FoldingRange[]> {
+    const ranges = await vscode.commands.executeCommand<vscode.FoldingRange[]>(
+        'vscode.executeFoldingRangeProvider',
+        uri
+    )
+    switch (type) {
+        case 'imports':
+            return ranges.filter(r => r.kind === vscode.FoldingRangeKind.Imports)
+        case 'comment':
+            return ranges.filter(r => r.kind === vscode.FoldingRangeKind.Comment)
+        case 'regions':
+            return ranges.filter(
+                r => r.kind !== vscode.FoldingRangeKind.Imports && r.kind !== vscode.FoldingRangeKind.Comment
+            )
+        default:
+            return ranges
+    }
+}
+
+/**
+ * Gets the folding range containing the active cursor position.
+ *
+ * @param uri - The URI of the document to get folding ranges for
+ * @param activeCursor - The active cursor position
+ * @returns The folding range containing the active cursor, or undefined if none found
+ */
+export async function getCursorFoldingRange(
+    uri: vscode.Uri,
+    activeCursor: number
+): Promise<vscode.Selection | undefined> {
+    const symbols = await getSymbolRanges(uri).then(r => r.filter(s => s.kind === vscode.SymbolKind.Class))
+    // Get the folding ranges for the document,
+    const ranges = await getFoldingRange(uri).then(r => r.filter(r => r.kind !== 2 && r.kind !== 1 && r.kind !== 3))
```

Comment:
```suggestion
    const [symbols, ranges] = await Promise.all([
    		getSymbolRanges(uri).then(r => r.filter(s => s.kind === vscode.SymbolKind.Class)),
    		// Get the folding ranges for the document,
    		getFoldingRange(uri).then(r => r.filter(r => r.kind !== 2 && r.kind !== 1 && r.kind !== 3))
    ]
```

Can be parallelized 😎

Reply:
What do the magic numbers mean (kind 2, 1, or 3)? Is there an enum we can use instead?

Reply:
Yep the enums are available and can be used they actually make it easier to understand `vscode.SymbolKind.Module`, `vscode.SymbolKind.Namespace` and `vscode.SymbolKind.Package`

Reply:
oooo good catch. I just replaced it with !r.kind since we just want ranges that don't belong to any of the folding range kind!